pkg/providers/v1/instances_v2.go

```diff
@@ -0,0 +1,81 @@
+// This file implements the InstancesV2 interface.
+// InstancesV2 is an abstract, pluggable interface for cloud provider instances.
+// Unlike the Instances interface, it is designed for external cloud providers and should only be used by them.
+
+package aws
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+// InstancesV2.InstanceExists
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+func (c *Cloud) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	return c.InstanceExistsByProviderID(ctx, providerID)
+}
+
+// InstancesV2.InstanceShutdown
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+func (c *Cloud) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	return c.InstanceShutdownByProviderID(ctx, providerID)
+}
+
+// InstancesV2.InstanceMetadata
+// InstanceMetadata returns the instance's metadata. The values returned in InstanceMetadata are
+// translated into specific fields and labels in the Node object on registration.
+// Implementations should always check node.spec.providerID first when trying to discover the instance
+// for a given node. In cases where node.spec.providerID is empty, implementations can use other
+// properties of the node like its name, labels and annotations.
+func (c *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return &cloudprovider.InstanceMetadata{}, err
```

Comment:
is `return nil, err` enough?

Reply:
Maybe I'm wrong, but I thought you needed to return an empty object with Go... I might also be conflating with another language.

Reply:
Sometimes people use signatures that return `(struct, error)` instead of `(*struct, error)`. The former forces you to return an empty struct, while the latter lets you return `nil, err`.

Reply:
Tx for being big brained @cartermckinnon